src/common/hist_util.cc

```diff
@@ -524,31 +653,26 @@ FindGroups(const std::vector<unsigned>& feature_list,
       search_groups.resize(param.max_search_group);
     }
 
-    // examine each candidate group: is it okay to insert fid?
-    for (auto gid : search_groups) {
-      const size_t rest_max_cnt = max_conflict_cnt - group_conflict_cnt[gid];
-      const size_t cnt = GetConflictCount(conflict_marks[gid], column, rest_max_cnt);
-      if (cnt <= rest_max_cnt) {
-        need_new_group = false;
-        groups[gid].push_back(fid);
-        group_conflict_cnt[gid] += cnt;
-        group_nnz[gid] += cur_fid_nnz - cnt;
-        MarkUsed(&conflict_marks[gid], column);
+    switch (colmat.GetTypeSize()) {
+      case sizeof(uint8_t):
+        SetGroup(fid, colmat.GetColumn<uint8_t>(fid), max_conflict_cnt, search_groups,
+                 &group_conflict_cnt, &conflict_marks, &groups, &group_nnz, cur_fid_nnz, nrow);
         break;
-      }
-    }
-
-    // create new group if necessary
-    if (need_new_group) {
-      groups.emplace_back();
-      groups.back().push_back(fid);
-      group_conflict_cnt.push_back(0);
-      conflict_marks.emplace_back(nrow, false);
-      MarkUsed(&conflict_marks.back(), column);
-      group_nnz.emplace_back(cur_fid_nnz);
+      case sizeof(uint16_t):
+        SetGroup(fid, colmat.GetColumn<uint16_t>(fid), max_conflict_cnt, search_groups,
+                 &group_conflict_cnt, &conflict_marks, &groups, &group_nnz, cur_fid_nnz, nrow);
+        break;
+      case sizeof(uint32_t):
+        SetGroup(fid, colmat.GetColumn<uint32_t>(fid), max_conflict_cnt, search_groups,
+                 &group_conflict_cnt, &conflict_marks, &groups, &group_nnz, cur_fid_nnz, nrow);
+        break;
+      default:
+        size_t curent_type_size = colmat.GetTypeSize();
+        CHECK(curent_type_size == UINT8_BINS_TYPE ||
+              curent_type_size == UINT16_BINS_TYPE ||
+              curent_type_size == UINT32_BINS_TYPE);
```

Comment:
ASSERT is enough

Reply:
CHECK(false) was used in all places instead, as I didn't find any other ASSERT usages